packages/llmRunner/src/llmRunner/LLMStep.ts

```diff
@@ -0,0 +1,235 @@
+import { SqError, SqProject } from "@quri/squiggle-lang";
+
+import { Artifact, ArtifactKind } from "./Artifact";
+import { LlmMetrics, Message } from "./LLMClient";
+import { LogEntry, Logger, TimestampedLogEntry } from "./Logger";
+import { PromptPair } from "./prompts";
+import { Workflow } from "./Workflow";
+
+export type CodeState =
+  | {
+      type: "formattingFailed";
+      error: string;
+      code: string;
+    }
+  | { type: "runFailed"; code: string; error: SqError; project: SqProject }
+  | { type: "success"; code: string };
+
+export function codeStateErrorString(codeState: CodeState): string {
+  if (codeState.type === "formattingFailed") {
+    return codeState.error;
+  } else if (codeState.type === "runFailed") {
+    return codeState.error.toStringWithDetails();
+  }
+  return "";
+}
+
+export type StepShape<
+  I extends Record<string, ArtifactKind> = Record<string, ArtifactKind>,
+  O extends Record<string, ArtifactKind> = Record<string, ArtifactKind>,
+> = {
+  inputs: I;
+  outputs: O;
+};
+
+type ExecuteContext<Shape extends StepShape> = {
```

Comment:
I intentionally decided to use `execute(context)` callbacks instead of inheritance — I predict inheritance would get messy if we did something like `Step` -> `LLMStep` -> `FixSquiggleLLMStep` hierarchy. Functions are easier to compose without inheritance.

I also pass `ExecuteContext` object instead of passing the entire `LLMStepInstance` to limit the API surface for execute functions — they shouldn't be able to call instance internals.